sled-agent/src/services.rs

```diff
@@ -423,88 +576,14 @@ impl ServiceManager {
-            );
-            let cfg: AllZoneRequests = toml::from_str(
-                &tokio::fs::read_to_string(&config_path)
-                    .await
-                    .map_err(|err| Error::io_path(&config_path, err))?,
-            )
-            .map_err(|err| Error::TomlDeserialize {
-                path: config_path.clone(),
-                err,
-            })?;
-            let mut existing_zones = self.inner.zones.lock().await;
-
-            // Initialize and DNS and NTP services first as they are required
-            // for time synchronization, which is a pre-requisite for the other
-            // services.
-            self.initialize_services_locked(
-                &mut existing_zones,
-                &cfg.requests
-                    .clone()
-                    .into_iter()
-                    .filter(|svc| {
-                        matches!(
-                            svc.zone.zone_type,
-                            ZoneType::InternalDns | ZoneType::Ntp
-                        )
-                    })
-                    .collect(),
-            )
-            .await?;
-
-            drop(existing_zones);
-
-            info!(&self.inner.log, "Waiting for sled time synchronization");
-
-            retry_notify(
-                retry_policy_local(),
-                || async {
-                    match self.timesync_get().await {
-                        Ok(TimeSync { sync: true, .. }) => {
-                            info!(&self.inner.log, "Time is synchronized");
-                            Ok(())
-                        }
-                        Ok(ts) => Err(BackoffError::transient(format!(
-                            "No sync {:?}",
-                            ts
-                        ))),
-                        Err(e) => Err(BackoffError::transient(format!(
-                            "Error checking for time synchronization: {}",
-                            e
-                        ))),
-                    }
-                },
-                |error, delay| {
-                    warn!(
-                        self.inner.log,
-                        "Time not yet synchronised (retrying in {:?})",
-                        delay;
-                        "error" => ?error
-                    );
-                },
-            )
-            .await
-            .expect("Expected an infinite retry loop syncing time");
-
-            let mut existing_zones = self.inner.zones.lock().await;
-
-            // Initialize all remaining serivces
-            self.initialize_services_locked(&mut existing_zones, &cfg.requests)
-                .await?;
-        } else {
-            info!(
-                &self.inner.log,
-                "No sled services found at {}",
-                config_path.to_string_lossy()
-            );
-        }
+        self.load_non_storage_services().await?;
+        // TODO: These will fail if the disks aren't attached.
```

Comment:
I'm having a hard time reasoning about when we'd want to retry, since it's unclear to me when a disk would become attached within a retry period. My admittedly, somewhat uninformed, take at this moment is that we shouldn't retry. 

I think the distinction between these calls not retrying and NTP retrying is twofold:
 1. NTP retries even on successful startup, since we are waiting for time sync
 2. NTP is a barrier to starting other services. If it fails later services will not work.

Reply:
This definitely deserves a follow-up bug - it's only relevant in the "cold boot" case, but that does matter!

Here's the deal:

- When we boot the sled agent, it's possible that not all disks have been parsed (e.g., suppose there's a U.2 that's slow to bind a driver).
- When we call this function, we'll read the service ledger from the M.2s, see what services *should* be started, and try starting them.
- If any of those services...
  - ... have datasets in the U.2s
  - ... have zone filesystems in the U.2s
- ... then we'd fail to start them.

But that doesn't mean we should *never* launch the service - if it's in the M.2 service ledger, either RSS or Nexus provisioned that service, so we should keep giving it a shot. "The driver did bind, but it just took a while"  and "The U.2 was unplugged, but someone put it back" are both cases where we should be able to launch these services, even if we might initially fail.

If, after a long enough period of time, we determine that we cannot launch that service (if the U.2 is detached, fails, etc), then we'd have an opportunity to do some notification through the fault tolerance system (we either tell Nexus that the service isn't booting, or Nexus notices on its own somehow).

Reply:
Filed https://github.com/oxidecomputer/omicron/issues/2973 , will mention it in this PR

Reply:
Mentioned in https://github.com/oxidecomputer/omicron/pull/2946/commits/ed20fffa842f9f048373d59cff5fe157794b1e6f

Reply:
Thanks for the explanation. Those justifications make sense to me.